rust/src/remove_duplicates_list.rs

```diff
@@ -0,0 +1,58 @@
+use crate::Solution;
+
+// Definition for singly-linked list.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct ListNode {
+    pub val: i32,
+    pub next: Option<Box<ListNode>>,
+}
+
+impl ListNode {
+    fn from(ar: Vec<i32>) -> Option<Box<ListNode>> {
+        ar.iter().rev().fold(None, |acc, i| {
+            Option::Some(Box::new(ListNode { val: *i, next: acc }))
+        })
+    }
+
+    fn to_vec(self) -> Vec<i32> {
+        let mut res = Vec::new();
+        let mut ptr = Some(Box::new(self));
+        while let Some(node) = ptr {
+            res.push(node.val);
+            ptr = node.next;
+        }
+        return res;
+    }
+}
+
+impl Solution {
+    pub fn delete_duplicates(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
```

Comment:
For me the recursive nature of this task suggests itself to improve the elegance, but it may not be that obvious here. But Rust is not a functional language whatever people would say,  it doesn't have Tail Call Optimization and overall not very convenient for recursive functions. But I don't really care :) Here the first recursive solution comes to my mind.

```
fn delete_duplicates(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        fn delete_first_occurrence_of(val: i32, head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
            match head {
                 Some(node) if node.val == val => delete_first_occurrence_of(val, node.next),
                 _ => head
            }
        }
        if let Some(mut first) = head {
              first.next = Self::delete_duplicates(delete_first_occurrence_of(first.val, first.next));
              Some(first)
        } else {
              None
        }
    }
```

Reply:
Yes, input is sorted by task.